shared_model/builders/protobuf/block.hpp

```diff
@@ -0,0 +1,123 @@
+ */
+
+#ifndef IROHA_PROTO_BLOCK_BUILDER_HPP
+#define IROHA_PROTO_BLOCK_BUILDER_HPP
+
+#include "backend/protobuf/block.hpp"
+#include "backend/protobuf/queries/proto_query.hpp"
+#include "backend/protobuf/transaction.hpp"
+#include "block.pb.h"
+
+#include "builders/protobuf/unsigned_proto.hpp"
+#include "interfaces/base/hashable.hpp"
+#include "interfaces/base/signable.hpp"
+#include "interfaces/common_objects/types.hpp"
+#include "interfaces/iroha_internal/block.hpp"
+#include "interfaces/transaction.hpp"
+
+namespace shared_model {
+  namespace proto {
+    template <int S = 0>
+    class TemplateBlockBuilder {
+     private:
+      template <class T>
+      using w = detail::PolymorphicWrapper<T>;
+
+      template <int>
+      friend class TemplateBlockBuilder;
+
+      enum RequiredFields {
+        Transactions,
+        TxNumber,
+        Height,
+        PrevHash,
+        CreatedTime,
+        Signatures,
+        TOTAL
+      };
+
+      template <int s>
+      using NextBuilder = TemplateBlockBuilder<S | (1 << s)>;
+
+      iroha::protocol::Block block_;
+
+      template <int Sp>
+      TemplateBlockBuilder(const TemplateBlockBuilder<Sp> &o)
+          : block_(o.block_) {}
+
+     public:
+      TemplateBlockBuilder() = default;
+
+      NextBuilder<Transactions> transactions(
+          const std::vector<w<Transaction>> &transactions) {
+        for (auto tx : transactions) {
+          iroha::protocol::Transaction proto_tx(tx->getTransport());
+          block_.mutable_payload()->mutable_transactions()->Add(std::move(proto_tx));
+        }
+        return *this;
+      }
+
+      NextBuilder<TxNumber> tx_number(
+          interface::Block::TransactionsNumberType tx_number) {
+        block_.mutable_payload()->set_tx_number(tx_number);
+        return *this;
+      }
+
+      NextBuilder<Height> height(interface::types::HeightType height) {
+        block_.mutable_payload()->set_height(height);
+        return *this;
+      }
+
+      NextBuilder<PrevHash> prev_hash(crypto::Hash hash) {
+        block_.mutable_payload()->set_prev_block_hash(
+            crypto::toBinaryString(hash));
+        return *this;
+      }
+
+      NextBuilder<CreatedTime> created_time(
+          interface::types::TimestampType time) {
+        block_.mutable_payload()->set_created_time(time);
+        return *this;
+      }
+
+      NextBuilder<Signatures> signatures(
+          const interface::Signable<interface::Block,
+                                    iroha::model::Block>::SignatureSetType
+              &signatures) {
+        for (const auto &signature : signatures) {
+          auto sig = block_.add_signatures();
+
+          sig->set_pubkey(crypto::toBinaryString(signature->publicKey()));
+          sig->set_signature(crypto::toBinaryString(signature->signedData()));
+        }
+        return *this;
+      }
+
+      UnsignedWrapper<Block> build() {
+        static_assert(S == (1 << TOTAL) - 1, "Required fields are not set");
```

Comment:
I think we should add stateless validation here or a TODO note if it's subject of another task.